### ISSUE TITLE
Investigate and implement CRINGE loss

### ISSUE THREAD

**Issue by 0x000011b** [PygmalionAI/accelerate-training-code#2]
Meta has a paper called "The CRINGE Loss: Learning what language not to model", which claims to be able to alleviate problems like repetitions, contradictions and bad generations simply by training on a small set of negative examples using a different loss function.

Basic details can be found [here](https://parl.ai/projects/cringe/), PyTorch implementation is [here](https://github.com/facebookresearch/ParlAI/blob/main/projects/cringe/cringe_loss.py). It'd be awesome if we could get this integrated and give it a shot for penalizing short generations, contradictions and so on.

**Comment by TearGosling** [comment]
Will indeed work on this, already have a branch on the repo going.

**Comment by TearGosling** [comment]
I've opened PR #4 for implementing CRINGE loss. Marking this task as "under review".

**Comment by TearGosling** [comment]
Update on it. I've tested fine-tuning Pythia's 410M model on Anthropic's [HH-RLHF](https://huggingface.co/datasets/Anthropic/hh-rlhf) dataset with the CRINGE loss, representing the 'chosen' answer as positive tokens while representing the 'rejected' answer as negative tokens. Unfortunately, it seems that the model with CRINGE loss is a lot more incoherent compared to a standard fine-tune on only positive examples. I'm not sure of the exact cause of this, but it could be for a few reasons:

- I'm not treating examples right in some way. I will note that I had to modify the code a bit for CRINGE to have the loss work correctly. It is possible that I've misinterpreted something about the loss, and that classifier labels have to be treated differently (perhaps on a per-token basis rather than a per-word basis?) but I'm not sure on how to fix that.
- The CRINGE loss is not meant to be used like how I'm using it. Related to my previous concern, it may be the case that we're barking up the wrong tree by trying to have entire answers be classified as positive or negative rather than specific tokens, which is how the research paper seems to be treating it (?). We would either need to find a way to make certain tokens negative or modify CRINGE entirely to try and suit our purposes better.
- ParlAI's code is known to be broken and atrocious. It's quite possible that something in ParlAI's implementation of the loss is not working as it should be, and I'd need to track down exactly what that issue is and to fix it.
- Different hyperparameters may be required for a model using CRINGE than a model using regular cross-entropy loss. This shouldn't be the case, but considering that the training objective (and numbers being calculated) are indeed different from standard cross-entropy loss, it's possible that changes to these hyperparameters may result in an improvement in performance.

I'm not entirely sure what to do next at the moment. Suggestions would be very helpful, thank you.

**Comment by 0x000011b** [comment]
Thanks for the write-up. Regarding the first two points, I quote the paper's introduction:

> In this work, we study the setting where the training set involves a given set of positive example sequences, as is commonly used for language model training, and a set of negative example sequences, which are completions given a prompt that a model should not generate.

Given that they talk about "example sequences", I do think that we're on the right track by labeling entire completions as positive/negative instead of individual tokens.

Regarding this:

> - ParlAI's code is known to be broken and atrocious. It's quite possible that something in ParlAI's implementation of the loss is not working as it should be, and I'd need to track down exactly what that issue is and to fix it.

Indeed ParlAI has plenty of broken code. In my experience though, the vast majority of breakage there was thanks to their overuse of inheritance without proper testing for the relevant child classes, so when they'd modify some ancestor, stuff would break.

The CRINGE loss project is self-contained and hasn't seen any meaningful changes ever since it was initially committed though, so while it _is_ possible that there's something wrong with the implementation there, I'd initially treat that as pretty far down the likelihood scale.

I think these:

> - I will note that I had to modify the code a bit for CRINGE to have the loss work correctly. It is possible that I've misinterpreted something about the loss [...]

> - Different hyperparameters may be required for a model using CRINGE than a model using regular cross-entropy loss.

And possible data problems are the most probable culprits. If I were debugging this myself, I would:

- Create a dataset comprised exclusively of negative examples
- Take the model that was fine-tuned exclusively on the positive examples
- Fine-tune it on the negative examples. Do various runs with different hyperparameters.

That would hopefully allow us to rule in/out the hyperparameter theory. If that doesn't pan out, next up I'd say that possibly we either have a problematic dataset, or 410M is too small to accurately model whatever it is that differentiates the positive from the negative examples in that dataset.

In that case, I'd try replacing the HH data with something simpler, possibly synthetically generated. E.g.:
- Take some dataset that we can easily split apart based on the response labels (for example, a dataset where responses are always "yes" or "no", "entailment" and "contradiction" and so on)
- Train a base model without CRINGE (so only positive examples, meaning only correctly labeled data)
- Using the splits, generate some synthetic data to be used as negative examples and then fine-tune the base model using CRINGE
- Compare generation results of the base vs. fine-tuned model.

Theoretically you could try to do the "right" thing and invert labels to feed in as negative training examples (so end result would be a higher accuracy after fine-tuning with CRINGE) but this might be a bit too complex to work out on 410M. If you want, you could also/instead try just having all the negative examples use the same label (e.g. "no", "contradiction") and see if the model leans towards "yes" or "entailment" more often, which would imply CRINGE _seems_ to be working correctly and penalizing negative examples.

**Comment by 0x000011b** [comment]
Closing this one - after more internal tests this didn't seem like a promising way forward, so we're abandoning the idea for now.
